### ISSUE TITLE
Submission: 242 params, 99.07% accuracy — Spiral Addition Transformer

### ISSUE THREAD

**Issue by JackCai1206** [anadim/AdderBoard#33]
## Author
Jack Cai

## Unique Parameter Count
242

## Accuracy
**99.07%** (9917/10010 on official AdderBoard verification: 10 edge cases + 10,000 random pairs, seed=2025)

## Verification Output
```
Model: Spiral Addition Transformer (RMSNorm, shared-XYZ, tied-proj, FFN2+WD)
Author: Jack Cai
Parameters (unique): 242
Architecture: 1L decoder, d=6 (3 tok + 3 pos), 2h, hd=3, ff=2, RMSNorm, shared XYZ pos, tied output

Results: 9917/10010 correct (99.07%)
Time: 71.4s (140 additions/sec)
Status: QUALIFIED (threshold: 99%)
```

## Method

1-layer decoder-only transformer with a split-dimension architecture. `d_model=6` is split into two 3D subspaces: token values and positional information, concatenated (not summed).

### Architecture
- **d_model=6** = tok_dim(3) + pos_dim(3), concatenated
- **2 attention heads**, head_dim=3
- **FFN inner dim=2** (minimal, enabled by weight decay)
- **Split-head attention**: Q,K from positional dims; V from token dims
- **Tied output head**: `Linear(6,3) @ tok_emb.T` instead of `Linear(6,14)`
- **Shared XYZ positions**: X[i]=Y[i]=Z[i]=digit_enc[i]
- **RMSNorm** (no bias, no mean centering — saves 18 params vs LayerNorm)
- **LSB-first** digit ordering

### Key Tricks
1. **Spiral-initialized embeddings**: `(cos(2πd/10), sin(2πd/10), d/9)` for digits 0-9
2. **Split-head attention**: Q,K attend by position, V reads token values
3. **Shared XYZ positional encoding**: Same learned vectors for X, Y, Z operand positions (-60 params)
4. **Tied output head**: Project to 3D token subspace, dot with embedding table (-66 params)
5. **RMSNorm**: No bias/mean centering (-18 params vs LayerNorm)
6. **FFN dim=2 + weight decay 0.01**: Extreme compression via grokking
7. **LSB-first ordering**: Natural carry propagation

### Training
- Optimizer: AdamW (lr=1e-3, weight_decay=0.01)
- Batch size: 512, 50K samples/epoch
- Convergence: ~1000 epochs via grokking
- Hardware: Single CPU

## Link to Code
https://github.com/JackCai1206/smallest-addition-transformer

- Submission file: [`submission_242.py`](https://github.com/JackCai1206/smallest-addition-transformer/blob/master/submission_242.py)
- Checkpoint: [`best_model_242.pt`](https://github.com/JackCai1206/smallest-addition-transformer/blob/master/best_model_242.pt)
- Training script: [`experiments/scripts/train_spos_wd.py`](https://github.com/JackCai1206/smallest-addition-transformer/blob/master/experiments/scripts/train_spos_wd.py) `--variant d6_f2_sxyz_rms_wd01`

## Additional Notes
- We also have a **248-param variant** (shared LayerNorm, 99.50%) and a **260-param variant** (100.00% accuracy, zero failures)
- All variants use the same core architecture; see repo for full experiment log
- Previous trained-weights leaderboard leader: 311 params at 99.999%
- Code written with assistance from Claude Code
